### DISCUSSION TITLE
Comparison to -1 of unsigned integer.

### DISCUSSION THREAD

**Discussion by ghost** [raysan5/raylib#2054] [Q&A] [answered]
Hello. I am reading through the code and noticed the following branch condition:
https://github.com/raysan5/raylib/blob/2a6bd973bdfe92e9f8506feb766d3351c0802d94/src/rlgl.h#L2616

While `glInternalFormat` is defined [here](https://github.com/raysan5/raylib/blob/2a6bd973bdfe92e9f8506feb766d3351c0802d94/src/rlgl.h#L2611) as `unsigned int`. It should never be -1. Does this check work as intended, or does C perform some implicit conversion here?

**Comment by raysan5** [comment]
@hexaflex Actually, it looks like a bug! Good catch! Reviewed!

**Reply by ghost** [reply]
@raysan5  This seems to happen again at these locations:

* https://github.com/raysan5/raylib/blob/2a6bd973bdfe92e9f8506feb766d3351c0802d94/src/rlgl.h#L2764
* https://github.com/raysan5/raylib/blob/2a6bd973bdfe92e9f8506feb766d3351c0802d94/src/rlgl.h#L2836

**Comment by orcmid** [comment, marked as answer]
This looks like a tricky way to avoid `UINT_MAX` (from `limits.h`) when `glInternalFormat` might be a different unsigned size than `unsigned int` and the max value is some sort of flag case.

 I just compiled a small program using `cl.exe` with no options and that is exactly what happens.  I assigned `-1` to `unsigned int glInternalFormat` and the comparison `!= -1` failed and comparison with `UINT_MAX` succeeded.  

That's a beautiful hack.  A better solution might be to declare a `const` (better than `#define`) with the intended value and a big XXX comment with it, all in `rlgl.h`

**Reply by orcmid** [reply]
```c
/* umax.c 0.0.1                    UTF-8                      dh:2021-10-15
*---|----1----|----2----|----3----|----4----|----5----|----6----|----7----|--*

                     Check Unsigned Casting Trick
*/

#include <stdio.h>
#include <limits.h>

int main(void)
{
    unsigned int glInternalFormat;

    glInternalFormat = -1;

    if (glInternalFormat != -1)
         puts("-1 didn't work");
    else puts("-1 worked");
    if (glInternalFormat != UINT_MAX)
         puts("not UINT_MAX though");
    else puts("and it's UINT_MAX");
    }

/* 0.0.1 a simple check
   */
```

**Comment by orcmid** [comment]
![F21xy040-2021-10-15-1107-minusOne-umaxCheck](https://user-images.githubusercontent.com/3372580/137533654-e653f692-87e8-441a-9d45-ccb8e2305e9c.png)

**Comment by ghost** [comment]
Hah, interesting. Thanks for the input.

**Reply by orcmid** [reply]
It does depend on `unsigned int` data and operations being implemented as 2's-complement binary arithmetic.  It gets easy though.

The C99 Standard has `signed int` be converted to `unsigned int` in a comparison or other operation) with an `unsigned int`.  Assignment works the same way.  In the case of `unsigned int` values it is all modular arithmetic with modulus `UINT_MAX + 1`.  My copy of Harbison & Steele's _C A Reference Manual_ (ed.5) has this example on p.198,

```C
unsigned int UI = -1;
```

I would not  have seen that except this case in raylib had me look things up.

Funny.  Today I saw a thread with folks talking about undefined situations in C and C++.  This is clearly not one of them, as surprising as it might be.
